Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp

```diff
@@ -859,16 +859,17 @@ bool AppendPipeline::recycleTrackForPad(GstPad* demuxerSrcPad)
         return false;
     }
 
+    // Stopping parser/sink and restart them again (at the end) in order to support recycling in case the elements have
+    // received an EOS before having been unlinked.
```

Comment:
What is causing that EOS? Because that sounds like the root problem and this would only be a band-aid. For instance, if the demuxer sends an EOS, it is possible it won't accept new media segments, which is an even bigger problems.

Reply:
Qtdemux itself is causing the EOS on a "to be removed" stream before the AppendPipeline realizes that the stream is actually to be removed and tries to relink to a new demuxer pad.
This started to happen after this [downstream change](https://github.com/WebPlatformForEmbedded/buildroot/blob/main/package/gstreamer1/gst1-plugins-good/1.18.6/0015-qtdemux-emit-no-more-pads-after-pruning-old-pads.patch), which is actually [this upstream GStreamer change in qtdemux](https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/4535/diffs?commit_id=021572de93b364825955073af6493012aac375ac) (unfold all the source code lines to understand it), which comes from [this GStreamer MR](https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/4535).

What the GStreamer change does is to first "terminate" (signal EOS) old streams that are going to disappear, and then notify WebKit after that by sending the no-more-pads message. So, when WebKit detects no-more-pads, the old streams (and their parser and appsink elements) have already been EOSed.

Then WebKit tries to AppendPipeline::recycleTrackForPad(), that is, to reuse former parser and appsink elements from the old terminated stream. That's why it's needed to "restart" those EOSed elements.

What happens before qtdemux is irrelevant. Qtdemux itself is never receiving EOS in this case.

Reply:
> Qtdemux itself is never receiving EOS in this case.

For the record, it shouldn't, in general in AppendPipeline.

What you explain makes sense.

Reply:
Add that explanation to the comment above these lines of code. r+